third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,885 @@
+    cached_type_params: Vec<TypeParameter>,
+
+    // Tracks which definition (`Let` statement `NodeId`) is visible during visit to find modified
+    // local vars.  A use of a symbol which is missing must be a `Parameter`.  This is used only
+    // to determine if a symbol is in `unsafe_variables`.
+    visiting_binding: ScopedMap<Symbol, NodeId>,
+
+    // Unsafe variables are keyed by `Symbol` and `Let` statement `NodeId`,
+    // except function parameters, which are denoted by `None`.
+    unsafe_variables: BTreeSet<(Symbol, Option<NodeId>)>,
+
+    // Rename symbols to avoid name conflicts between nested symbols:
+    // The outermost occurrence of a `Symbol` is mapped to itself.  Other symbols whose
+    // scope is nested inside other identical symbols are mapped to new, unique symbols.
+    // (This is used to allow references to symbols which are shadowed by inner reuse.)
+    remapped_symbol: ScopedMap<Symbol, Symbol>,
+
+    // Tracks constant values from scope.  All symbols in keys and values are remapped.  A variable
+    // without a constant value in this scope is set to `SimpleValue::NonConstant`, while a
+    // (remapped) variable `sym` with an unknown or unrepresentable value is bound to `LocalVar(sym)`.
+    values: ScopedMap<Symbol, SimpleValue>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum SimpleValue {
+    Value(Value),
+    Temporary(TempIndex),
+    LocalVar(Symbol),
+    NonConstant,
+    Uninitialized,
+}
+
+impl<'env> SimplifierRewriter<'env> {
+    fn new(env: &'env GlobalEnv, eliminate_code: bool) -> Self {
+        let debug = env
+            .get_extension::<Options>()
+            .expect("Options is available")
+            .debug;
+        let constant_folder = ConstantFolder::new(env, false);
+        Self {
+            env,
+            debug,
+            constant_folder,
+            eliminate_code,
+            cached_params: Vec::new(),
+            cached_type_params: Vec::new(),
+            visiting_binding: ScopedMap::new(),
+            unsafe_variables: BTreeSet::new(),
+            remapped_symbol: ScopedMap::new(),
+            values: ScopedMap::new(),
+        }
+    }
+
+    /// Process a function.
+    pub fn rewrite_function_body(
+        &mut self,
+        params: &[Parameter],
+        type_params: &[TypeParameter],
+        exp: Exp,
+    ) -> Exp {
+        self.cached_params = params.to_vec();
+        self.cached_type_params = type_params.to_vec();
+        self.unsafe_variables = find_possibly_modified_vars(self.env, params, exp.as_ref());
+        self.visiting_binding.clear();
+        self.values.clear();
+        self.remapped_symbol.clear();
+        if SIMPLIFIER_DEBUG {
+            eprintln!("Unsafe variables are {:#?}", self.unsafe_variables);
+        }
+        // Enter Function scope (a specialized `rewrite_enter_scope()` call)
+        self.values.enter_scope();
+        self.remapped_symbol.enter_scope();
```

Comment:
Document why we don't need to `enter_scope()` for `self.visiting_binding`. Have not looked through carefully, but because of the `clear()`s above, these `enter_scope()` may not be necessary.